src/sources/pulsar.rs

```diff
@@ -0,0 +1,624 @@
+        }
+        if let Some(subscription_name) = &self.subscription_name {
+            consumer_builder = consumer_builder.with_subscription(subscription_name);
+        }
+
+        let consumer = consumer_builder.build::<String>().await?;
+
+        Ok(consumer)
+    }
+}
+
+async fn pulsar_source(
+    mut consumer: Consumer<String, TokioExecutor>,
+    decoder: Decoder,
+    mut shutdown: ShutdownSignal,
+    mut out: SourceSender,
+    acknowledgements: bool,
+    log_namespace: LogNamespace,
+) -> Result<(), ()> {
+    let (finalizer, mut ack_stream) =
+        OrderedFinalizer::<FinalizerEntry>::maybe_new(acknowledgements, Some(shutdown.clone()));
+
+    let bytes_received = register!(BytesReceived::from(Protocol::TCP));
+    let events_received = register!(EventsReceived);
+    let pulsar_error_events = register!(PulsarErrorEvent);
+
+    loop {
+        tokio::select! {
+            _ = &mut shutdown => break,
+            entry = ack_stream.next() => {
+                if let Some((status, entry)) = entry {
+                    handle_ack(&mut consumer, status, entry, &pulsar_error_events).await;
+                }
+            },
+            Some(maybe_message) = consumer.next() => {
+                match maybe_message {
+                    Ok(msg) => {
+                        bytes_received.emit(ByteSize(msg.payload.data.len()));
+                        parse_message(msg, &decoder, &finalizer, &mut out, &mut consumer, log_namespace, &events_received, &pulsar_error_events).await;
+                    }
+                    Err(error) => {
+                        pulsar_error_events.emit(PulsarErrorEventData{
+                            msg: error.to_string(),
+                            error_type:PulsarErrorEventType::Read,
+                        });
+                    }
+                }
+            },
+        }
+    }
+
+    Ok(())
+}
+
+#[allow(clippy::too_many_arguments)]
+async fn parse_message(
+    msg: Message<String>,
+    decoder: &Decoder,
+    finalizer: &Option<OrderedFinalizer<FinalizerEntry>>,
+    out: &mut SourceSender,
+    consumer: &mut Consumer<String, TokioExecutor>,
+    log_namespace: LogNamespace,
+    events_received: &Registered<EventsReceived>,
+    pulsar_error_events: &Registered<PulsarErrorEvent>,
+) {
+    let publish_time = i64::try_from(msg.payload.metadata.publish_time)
+        .ok()
+        .and_then(|millis| chrono::Utc.timestamp_millis_opt(millis).latest());
+    let topic = msg.topic.clone();
+    let producer_name = msg.payload.metadata.producer_name.clone();
+
+    let mut stream = FramedRead::new(msg.payload.data.as_ref(), decoder.clone());
+    let stream = async_stream::stream! {
+        while let Some(next) = stream.next().await {
+            match next {
+                Ok((events, _byte_size)) => {
+                    events_received.emit(CountByteSize(
+                        events.len(),
+                        events.size_of().into(),
```

Comment:
Why do we ignore the `byte_size` here? Also, should we be using `events.estimated_json_encoded_size_of()` instead?

Reply:
Good catch!